### PR TITLE
Add gio_futures example

### PR DESCRIPTION
**Pull request by sdroege**
Depends on https://github.com/gtk-rs/gio/pull/108

I'd also like to add an example using async/await but that would require nightly. How would we handle that?

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
By adding a feature called "nighty-only" or something along the lines? I restarted CIs.

File: Cargo.toml [hunk 1]
```diff
@@ -36,3 +40,80 @@ gtk_3_16 = ["gtk_3_10", "gtk/v3_16"]
 gtk_3_18 = ["gtk_3_16", "gtk/v3_18"] #for CI tools
 gtk_3_20 = ["gtk_3_18", "gtk/v3_20"] #for CI tools
 gtk_3_22 = ["gtk_3_20", "gtk/v3_22"] #for CI tools
+futures = ["futures-core", "futures-executor", "futures-channel", "futures-util", "glib/futures", "gio/futures"]
+
+[patch.'https://github.com/gtk-rs/gio']
+gio = { git = 'https://github.com/sdroege/gio-rs', branch = 'futures' }
```
Comment (sdroege) [inline Cargo.toml]: This currently patches in my GIO branch from https://github.com/gtk-rs/gio/pull/108. Will remove once that is merged

File: Cargo.toml [hunk 2]
```diff
@@ -36,3 +40,80 @@ gtk_3_16 = ["gtk_3_10", "gtk/v3_16"]
...
+[[bin]]
+name = "text_viewer"
+
+[[bin]]
+name = "transparent_main_window"
+
+[[bin]]
+name = "treeview"
+
+[[bin]]
```
Comment (EPashkin) [inline Cargo.toml]: Strange, why [[bin]] added?

Comment (EPashkin) [reply Cargo.toml]: Oh, understand

Comment (EPashkin) [reply Cargo.toml]: Maybe better do same as we do with version?
https://github.com/gtk-rs/examples/blob/master/src/bin/builder_basics.rs#L76-L85

File: Cargo.toml [hunk 3]
```diff
@@ -36,3 +40,80 @@ gtk_3_16 = ["gtk_3_10", "gtk/v3_16"]
 gtk_3_18 = ["gtk_3_16", "gtk/v3_18"] #for CI tools
 gtk_3_20 = ["gtk_3_18", "gtk/v3_20"] #for CI tools
 gtk_3_22 = ["gtk_3_20", "gtk/v3_22"] #for CI tools
+futures = ["futures-core", "futures-executor", "futures-channel", "futures-util", "glib/futures", "gio/futures"]
```
Comment (EPashkin) [inline Cargo.toml]: Maybe here also must be some `gtk_3_xxx` ?

**Comment by EPashkin** [comment]
And IMHO https://github.com/gtk-rs/examples/blob/pending/build_travis.sh#L8 IMHO also need `futures`

File: Cargo.toml (hunk 3 above)
Comment (sdroege) [reply Cargo.toml]: No GTK is needed here

**Comment by sdroege** [comment]
> And IMHO https://github.com/gtk-rs/examples/blob/pending/build_travis.sh#L8 also need futures

Agreed

File: Cargo.toml (hunk 2 above)
Comment (sdroege) [reply Cargo.toml]: Why do in code what cargo can do for us? :) I would rather convert all the examples

File: Cargo.toml (hunk 3 above)
Comment (EPashkin) [reply Cargo.toml]: Ok

File: Cargo.toml (hunk 2 above)
Comment (EPashkin) [reply Cargo.toml]: Current version don't accepted by cargo as it don't have `path`.

**Comment by EPashkin** [comment]
Please, add futures to `#default = ["gtk_3_22"]` too, it helps greatly for local checking

File: Cargo.toml (hunk 3 above)
Comment (EPashkin) [reply Cargo.toml]: `"glib/futures"` IMHO unneeded here, as it included in `"gio/futures"`

File: Cargo.toml (hunk 2 above)
Comment (EPashkin) [reply Cargo.toml]: Oops, it just because `variant.rs` do not exists

Comment (sdroege) [reply Cargo.toml]: Hmm I have that file :) I'll check and fix later

**Comment by sdroege** [comment]
async-await example would also require using `futures-stable`, `Pin` and all that inside glib for the `Executor` and everything. I'd postpone that for some other time

**Comment by sdroege** [comment]
> async-await example would also require using futures-stable, Pin and all that inside glib for the Executor and everything. I'd postpone that for some other time

And with that I mean in the next days in separate PRs, not the too distant future. I have something half-working

**Comment by sdroege** [comment]
@GuillaumeGomez Let's get this in then, or is there anything missing still?

**Comment by GuillaumeGomez** [comment]
I was waiting for CIs and forgot about it. Sorry about that...

**Pull request merged**